ports/psoc6/machine_pdm_pcm.c

```diff
@@ -0,0 +1,355 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2022-2024 Infineon Technologies AG
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdlib.h>
+
+#include "py/runtime.h"
+#include "py/mphal.h"
+#include "machine_pin_phy.h"
+#include "modmachine.h"
+#include "mplogger.h"
+
+#define pdm_pcm_assert_raise_val(msg, ret)   if (ret != CY_RSLT_SUCCESS) { \
+        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT(msg), ret); \
+}
+
+#define DEFAULT_LEFT_GAIN       0
+#define DEFAULT_RIGHT_GAIN      0
+#define MICROPY_HW_MAX_PDM_PCM  1
+
+#define AUDIO_SYS_CLOCK_24_576_000_HZ      24576000u /* (Ideally 24.576 MHz) For sample rates: 8 KHz / 16 KHz / 48 KHz */
+#define AUDIO_SYS_CLOCK_22_579_000_HZ      22579000u /* (Ideally 22.579 MHz) For sample rates: 22.05 KHz / 44.1 KHz */
+
+cyhal_clock_t pdm_pcm_audio_clock;
+
+// Constructor args
+enum {
+    ARG_clk,
+    ARG_data,
+    ARG_sample_rate,
+    ARG_decimation_rate,
+    ARG_bits,
+    ARG_format,
+    ARG_left_gain,
+    ARG_right_gain,
+};
+
+typedef enum {
+    BITS_16 = 16,
+    BITS_18 = 18,
+    BITS_20 = 20,
+    BITS_24 = 24
+} pdm_pcm_word_length_t;
+
+typedef enum {
+    RX,
+    TX
```

Comment:
Not sure if it is worth to keep the potential TX mode... 
Not sure if other vendors have such PDM TX... 
I would remove it, because it is effectively enabling nothing, and can lead to confusion in code readability. In can be always added later.

Reply:
Done